jax/experimental/sparsify.py

```diff
@@ -0,0 +1,300 @@
+    closed_jaxpr = make_jaxpr(f)(*argspecs_to_avals(spenv, argspecs))
+    return eval_sparse(closed_jaxpr.jaxpr, closed_jaxpr.literals, argspecs, spenv)
+  return wrapped
+
+def sparsify(f):
+  f_raw = sparsify_raw(f)
+  @functools.wraps(f)
+  def wrapped(*args, **params):
+    spenv = SparseEnv()
+    argspecs = arrays_to_argspecs(spenv, args)
+    argspecs_out, multiple_results = f_raw(spenv, *argspecs, **params)
+    out = argspecs_to_arrays(spenv, argspecs_out)
+    return out if multiple_results else out[0]
+  return wrapped
+
+def _zero_preserving_unary_op(prim):
+  def func(spenv, *argspecs, **kwargs):
+    assert len(argspecs) == 1
+    buf = argspecs[0].data(spenv)
+    buf_out = prim.bind(buf, **kwargs)
+    out_argspec = ArgSpec(argspecs[0].shape, spenv.push(buf_out), argspecs[0].indices_ref)
+    return (out_argspec,)
+  return func
+
+for _prim in [
+    lax.abs_p, lax.expm1_p, lax.log1p_p, lax.neg_p, lax.sign_p, lax.sin_p,
+    lax.sinh_p, lax.sqrt_p, lax.tan_p, lax.tanh_p, lax.convert_element_type_p
+  ]:
+  sparse_rules[_prim] = _zero_preserving_unary_op(_prim)
+
+def _dot_general_sparse(spenv, *argspecs, dimension_numbers, precision, preferred_element_type):
+  assert len(argspecs) == 2
+  assert argspecs[0].is_sparse() and not argspecs[1].is_sparse()
+  args = argspecs_to_arrays(spenv, argspecs)
+
+  result = sparse_ops.bcoo_dot_general(args[0].data, args[0].indices, args[1],
+                                       lhs_shape=args[0].shape,
+                                       dimension_numbers=dimension_numbers)
+  argspec = ArgSpec(result.shape, spenv.push(result), None)
+  return [argspec]
+
+sparse_rules[lax.dot_general_p] = _dot_general_sparse
+
+def _transpose_sparse(spenv, *argspecs, permutation):
+  assert len(argspecs) == 1
+  shape = argspecs[0].shape
+  data = argspecs[0].data(spenv)
+  indices = argspecs[0].indices(spenv)
+  assert len(shape) == len(permutation) == len(set(permutation))
+  assert min(permutation) == 0
+  assert max(permutation) == len(shape) - 1
+  if data.ndim > 1:
+    raise NotImplementedError("sparse transpose with batch & dense dimensions.")
+  out_indices = indices[permutation, :]
+  out_shape = tuple(shape[i] for i in permutation)
+  argspec = ArgSpec(out_shape, argspecs[0].data_ref, spenv.push(out_indices))
+  return (argspec,)
+
+sparse_rules[lax.transpose_p] = _transpose_sparse
+
+def _add_sparse(spenv, *argspecs):
+  X, Y = argspecs
+  if X.is_sparse() and Y.is_sparse():
+    if X.shape != Y.shape:
+      raise NotImplementedError("Addition between sparse matrices of different shapes.")
+    if X.indices_ref == Y.indices_ref:
+      out_data = lax.add(X.data(spenv), Y.data(spenv))
+      out_argspec = ArgSpec(X.shape, spenv.push(out_data), X.indices_ref)
+    elif X.indices(spenv).ndim != Y.indices(spenv).ndim or X.data(spenv).ndim != Y.data(spenv).ndim:
+      raise NotImplementedError("Addition between sparse matrices with different batch/dense dimensions.")
+    else:
```

Comment:
Note to selves: This case's implementation begets a requirement for BCOO that might rule out other operations (like `sin`). For now, here's what it would look like if we were to allow it.